pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,390 @@
+			}
+			// Configure SRIOV as a secondary network interface
+			if err = pc.configureSriovAsSecondaryInterface(pod, netinfo, containerInfo, result); err != nil {
+				// SRIOV interface creation failed. Free allocated IP address
+				if ipamerr = ipam.DelIPAMSubnetAddress(cniconfig, cmdArgs); ipamerr != nil {
+					klog.ErrorS(err, "IPAM de-allocation failed: ", ipamerr)
+				}
+				return err
+			}
+		} else {
+			klog.ErrorS(err, "InterfaceType not supported for Pod ", klog.KObj(pod))
+		}
+
+	}
+	return nil
+}
+
+func (pc *PodController) configureSecondaryNetwork(pod *corev1.Pod, networklist []*SecondaryNetworkObject, podCNIInfo []*cnipodcache.CNIConfigInfo) error {
+
+	for _, netinfo := range networklist {
+		klog.InfoS("Secondary Network Information:", netinfo)
+		if len(netinfo.NetworkName) > 0 {
+			netDefCRD, err := pc.netAttachDefClient.NetworkAttachmentDefinitions(pod.Namespace).Get(context.TODO(), netinfo.NetworkName, metav1.GetOptions{})
+			if err != nil {
+				klog.ErrorS(err, "NetworkAttachmentDefinitions.Get failed", "pod", klog.KObj(pod))
+				// NetworkAttachmentDefinition not found at this time. Return error to re-queue and re-try.
+				return err
+			}
+			cniconfig, err := netdefutils.GetCNIConfig(netDefCRD, "")
+			if err != nil {
+				klog.ErrorS(err, "net-attach-def: CNI config spec read error %v", err)
+				// NetworkAttachmentDefinition Spec.Config parsing failed. return error to re-queue and re-try.
+				return err
+			}
+			// secondary network information retrieved from API server. Proceed to configure secondary interface now.
+			if err = pc.configureSecondaryInterface(pod, netinfo, podCNIInfo, cniconfig); err != nil {
+				// Interface creation failed. We will not retry secondary network configuration further.
+				klog.ErrorS(err, "Interface creation failed. Secondary network not configured on Pod ", klog.KObj(pod))
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
+func (pc *PodController) Run(stopCh <-chan struct{}) {
+	defer func() {
+		klog.InfoS("Shutting down", controllerName)
+		pc.queue.ShutDown()
+	}()
+
+	klog.InfoS("Starting ", controllerName)
+	go pc.podInformer.Run(stopCh)
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, pc.podInformer.HasSynced) {
+		return
+	}
+
+	for i := 0; i < numWorkers; i++ {
+		go wait.Until(pc.Worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func checkForPodSecondaryNetworkAttachement(pod *corev1.Pod) (string, bool) {
+	netObj, netObjExist := pod.GetAnnotations()[networkAttachDefAnnotationKey]
+	if netObjExist {
+		return netObj, true
+	} else {
+		return netObj, false
+	}
+}
+
+func parsePodSecondaryNetworkAnnotation(netObj string) ([]*SecondaryNetworkObject, error) {
+	var secNetwork []*SecondaryNetworkObject
+
+	if strings.IndexAny(netObj, "[{\"") >= 0 {
+		if err := json.Unmarshal([]byte(netObj), &secNetwork); err != nil {
+			klog.ErrorS(err, "parsePodSecondaryNetworkAnnotation: failed to parse Pod Network Attachment Selection Annotation JSON format ", err)
```

Comment:
typically we do not add an extra log message for an error that will be logged further up the stack, we just wrap the error with `fmt.Errorf` (to avoid "duplicate" logs)